src/item_uid.h

```diff
@@ -0,0 +1,42 @@
+#ifndef ITEM_UID_H
+#define ITEM_UID_H
+
+#include <limits>
+
+#include "json.h"
+
+class item_uid : public JsonSerializer, public JsonDeserializer
```

Comment:
So #15682 covers some issues I hadn't considered. I'm going to go with your suggestion in it to use an 8-bit header, 16-bit id for player and then 48-bits for the item id.

I'm not sure about how a factory design would work? If you could stub out an example and/or a pull to this PR?

As to the `item_uid` class it's implemented this way so that it can be added to `item` as an ordinary member without needing to add innumerable special cases for when the `item` class is copied. Such copies occur extensively throughout the code and would quickly result in duplicate values. Implementing it as a class guarantees this cannot occur so prevents a whole class of bugs at source. The alternative would a custom copy-constructor `item` which would be brittle to changes in that classes members or even less appealing custom logic at each point items could be copied.